doc/SAI-Proposal-HW-Resource.md

```diff
@@ -0,0 +1,178 @@
+- SAI_OBJECT_STAGE_INGRESS:
+  Resource belongs to ingress stage of HW pipeline
+- SAI_OBJECT_STAGE_EGRESS:
+  Resource belongs to egress stage of HW pipeline
+
+## Capability Query
+----------------
+New  query API is introduced for an attribute's stage. NOS can query the stage of an object's attribute for a given HW support. One of the following stage is returned as part the API
+- SAI_OBJECT_STAGE_BOTH: Attribute is common to ingress and egress stage
+- SAI_OBJECT_STAGE_INGRESS: Attribute is applicable only to ingress stage
+- SAI_OBJECT_STAGE_EGRESS: Attribute is applicabe only to egress stage
+
+Length of the returned array "stage" is attr_count. Caller must provide the buffer for array "stage".
+ 
+```
+sai_status_t sai_query_object_stage(
+        _In_ sai_object_id_t switch_id,
+        _In_ sai_object_type_t object_type,
+        _In_ uint32_t attr_count,
+        _In_ const sai_attribute_t *attr_list,
+        _Out_ sai_object_stage_t *stage);
+```
+
+## HOSTIF Trap Group
+-----------------
+A new CREATE_ONLY trap group attribute is introduced. NOS can set this attribute to specify the stage of the trap group.
+- SAI_HOTIF_TRAP_GROUP_ATTR_OBJECT_STAGE: 
+CREATE_ONLY attribute to configure a trap group of type stage ingress/egress/both
+
+## Policer Pool
+------------
+A new CREATE_ONLY policer attribute is introduced. NOS can set this attribute
+to specify the stage of the pool.
+- SAI_POLICER_ATTR_STAGE:
+  CREATE_ONLY attribute that specifies the stage of the policer pool. Default is kept as SAI_STAGE_BOTH common pool for backward compatibility.
+
+## Example
+-------
+Hostif trap's are implemented as an ingress or egress pipeline stage based on a HW architecture. In this example workflow, a hostif trap capability is queried. Based on returned value, appropriate trap group and policer pool is assigned to it. For e.g. a trap of type ingress will be assigned a trap group of type ingress using an ingress policer pool.
+
+## Sample Existing Workflow
+------------------------
+There is no concept of stages in this workflow.
+By default this trap is created in SAI_SWITCH_ATTR_DEFAULT_TRAP_GROUP trap group.
+
+```
+1. create_hostif_trap_group()
+2. create_hostif_trap(): 
+    By default this trap is created in 
+    SAI_SWITCH_ATTR_DEFAULT_TRAP_GROUP trap group.
+or
+2. create_hostif_trap():
+    For trap group created in step 1.
+3. create_policer()
+4. Attach policer to the trap using set of 
+    SAI_HOSTIF_TRAP_GROUP_ATTR_POLICER attribute.
+```
+
+### Sample New Workflow
+-------------------
+New workflow will identify if a given trap is applicable to ingress/egress/both. This will be done based on the enhanced capability query API.
+```
+1. Query the HW capability for a given trap
+    sai_attribute_t trap_attra[2];
+    trap_attra[0].id = 
+        SAI_HOSTIF_TRAP_ATTR_TRAP_TYPE;
+    trap_attrs[0].value.s32 = 
+        SAI_HOSTIF_TRAP_TYPE_DNAT_MISS;
+    
+    trap_attrs[2].id = 

```

Comment:
should this be trap_attrs[1]  instead of trap_attrs[2]?
reason is number of attributes is set to 2 in your query and you start with 0.

Reply:
Done
https://github.com/opencomputeproject/SAI/pull/1528/commits/5c17045bfd7cd33a14688dc25d32eaee88ab740a